File: packages/android/demo/src/main/kotlin/dev/crumb/demo/DemoCrumbConfiguration.kt
package dev.crumb.demo

import dev.crumb.core.CrumbConfiguration
import dev.crumb.core.CrumbDiagnosticsOptions
import dev.crumb.core.CrumbLogOptions
import dev.crumb.core.CrumbLogProvider
import dev.crumb.core.CrumbRelease
import dev.crumb.core.CrumbUploadOptions

internal data class DemoCrumbConfiguration(
    val crumb: CrumbConfiguration,
    val modeDescription: String,
) {
    companion object {
        fun make(
            projectKey: String = BuildConfig.CRUMB_DOGFOOD_PROJECT_KEY,
            ingestionUrl: String = BuildConfig.CRUMB_DOGFOOD_INGESTION_URL,
            environment: String = BuildConfig.CRUMB_DOGFOOD_ENVIRONMENT,
            logProvider: CrumbLogProvider? = null,
            appVersion: String = BuildConfig.VERSION_NAME,
            nativeBuild: String = BuildConfig.VERSION_CODE.toString(),
        ): DemoCrumbConfiguration {
            val resolvedKey = projectKey.trim().takeIf(String::isNotEmpty)
            val resolvedUrl = ingestionUrl.trim().trimEnd('/').takeIf(::isHttpUrl)
            val uploadEnabled = resolvedKey != null && resolvedUrl != null

            return DemoCrumbConfiguration(
                crumb = CrumbConfiguration(
                    projectKey = if (uploadEnabled) checkNotNull(resolvedKey) else "poc_write_key",
                    environment = if (uploadEnabled) environment.trim().ifEmpty { "staging" } else "local",
                    release = CrumbRelease(appVersion = appVersion, nativeBuild = nativeBuild),
                    diagnostics = CrumbDiagnosticsOptions(
                        healthCheckUrl = resolvedUrl?.let { "$it/health" }.takeIf { uploadEnabled },
                        logs = CrumbLogOptions(provider = logProvider),
                    ),
                    upload = CrumbUploadOptions(
                        ingestionUrl = resolvedUrl.takeIf { uploadEnabled },
                    ),
                ),
                modeDescription = if (uploadEnabled) "Staging upload enabled" else "Local-only mode",
            )
        }

        private fun isHttpUrl(value: String): Boolean = runCatching {
            val url = java.net.URI(value)
            (url.scheme == "https" || url.scheme == "http") && !url.host.isNullOrBlank()
        }.getOrDefault(false)
    }
}

File: examples/ios/CrumbDemo/DemoCrumbConfiguration.swift
import CrumbCore
import Foundation

struct DemoCrumbConfiguration {
    let crumb: CrumbConfiguration
    let modeDescription: String

    static func make(
        environment: [String: String] = ProcessInfo.processInfo.environment,
        bundle: Bundle = .main
    ) -> DemoCrumbConfiguration {
        let projectKey = resolvedValue(
            environment["CRUMB_DOGFOOD_PROJECT_KEY"],
            bundle.object(forInfoDictionaryKey: "CrumbDogfoodProjectKey") as? String
        )
        let ingestionValue = resolvedValue(
            environment["CRUMB_DOGFOOD_INGESTION_URL"],
            bundle.object(forInfoDictionaryKey: "CrumbDogfoodIngestionURL") as? String
        )
        let configuredEnvironment = resolvedValue(
            environment["CRUMB_DOGFOOD_ENVIRONMENT"],
            bundle.object(forInfoDictionaryKey: "CrumbDogfoodEnvironment") as? String
        )
        let ingestionURL = ingestionValue.flatMap(validHTTPURL)
        let uploadEnabled = projectKey != nil && ingestionURL != nil
        let appVersion = bundle.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String
        let nativeBuild = bundle.object(forInfoDictionaryKey: "CFBundleVersion") as? String

        return DemoCrumbConfiguration(
            crumb: CrumbConfiguration(
                projectKey: uploadEnabled ? (projectKey ?? "poc_write_key") : "poc_write_key",
                environment: uploadEnabled ? (configuredEnvironment ?? "staging") : "local",
                release: CrumbRelease(
                    appVersion: appVersion ?? "0.1.0",
                    nativeBuild: nativeBuild ?? "1"
                ),
                diagnostics: CrumbDiagnosticsOptions(
                    healthCheckURL: uploadEnabled ? ingestionURL?.appendingPathComponent("health") : nil
                ),
                upload: CrumbUploadOptions(ingestionURL: uploadEnabled ? ingestionURL : nil)
            ),
            modeDescription: uploadEnabled ? "Staging upload enabled" : "Local-only mode"
        )
    }

    private static func resolvedValue(_ candidates: String?...) -> String? {
        candidates.lazy.compactMap { candidate in
            guard let value = candidate?.trimmingCharacters(in: .whitespacesAndNewlines),
                  !value.isEmpty,
                  !value.contains("$(") else { return nil }
            return value
        }.first
    }

    private static func validHTTPURL(_ value: String) -> URL? {
        guard let url = URL(string: value),
              let scheme = url.scheme?.lowercased(),
              scheme == "https" || scheme == "http",
              url.host != nil else { return nil }
        return url
    }
}
